File: src/main/java/DigitalAssistant/skillLogic/FaceDetection.java
package DigitalAssistant.skillLogic;

import java.util.ArrayList;
import org.opencv.core.*;
import org.opencv.imgproc.Imgproc;
import org.opencv.videoio.VideoCapture;
import nu.pattern.OpenCV;
import org.opencv.objdetect.CascadeClassifier;
import org.opencv.core.MatOfRect;
import org.opencv.highgui.HighGui;
import org.opencv.imgcodecs.Imgcodecs;
import org.opencv.core.Mat;
import org.opencv.core.Rect;
import org.opencv.core.Scalar;


public class FaceDetection {

    public FaceDetection(){
        //faceDetection();
    }
    /**
     * Method runs face detection based on pre-trained Haar cascade classifier!
     * @return 
     */
    public boolean faceDetection() {
        try {
            ArrayList<Mat> our_images = new ArrayList<>();
            long startTime = System.currentTimeMillis();
    
            // Load the OpenCV native library
            OpenCV.loadLocally();

            // Open the default camera (index 0) 
            VideoCapture cap = new VideoCapture(0);

            // Check if camera opened successfully
            if (!cap.isOpened()) {
                System.out.println("Failed to open camera!");
                return false; // return false;
            }

            // Load the pre-trained Haar cascade classifier for face detection
            CascadeClassifier faceCascade = new CascadeClassifier();
            faceCascade.load("src/main/resources/DigitalAssistant/haarcascades/haarcascade_frontalface_alt.xml");

            boolean faceDetected = false;
            Mat frame = new Mat();
            Mat gray = new Mat();
            Mat faceROI = new Mat();
            String imagePath = "";

            // Loop to continuously capture frames
            while (true) {
                
                cap.read(frame);
                // DOWNSCALE the frame for faster processing
                // Mat downscaledFrame = new Mat();
                //Imgproc.resize(frame, downscaledFrame, new Size(frame.cols() / 1.2, frame.rows() / 1.2));

                // Convert the frame to grayscale for faster processing
                Imgproc.cvtColor(frame, gray, Imgproc.COLOR_BGR2GRAY);
                // basic processing for face detection
                // 1. denoising with gaussian filter

                Mat filteredImage = new Mat();
                // Apply Gaussian filter
                Imgproc.GaussianBlur(gray, filteredImage, new Size(5, 5), 0);
            
                // Detect faces in the frame using the Haar cascade classifier
                MatOfRect faces = new MatOfRect();
                //faceCascade.detectMultiScale(gray, faces, 1.3, 5);     // set parameters for face detection here
                faceCascade.detectMultiScale(filteredImage, faces, 1.5, 5);     

                // Check if any faces were detected
                if (!faces.empty()) {
                    faceDetected = true;
                    System.out.println("+++ Detected face +++");
                
                    for (Rect rect : faces.toArray()) {
                        Imgproc.rectangle(frame, rect.tl(), rect.br(), new Scalar(0, 255, 0), 5);   // Draw a rectangle around each face
                        faceROI = gray.submat(rect);
                    }

                    // Save the captured frame as an image file
                    imagePath = "src/main/resources/DigitalAssistant/haarcascades/detectedFace.jpg";
                    Imgcodecs.imwrite(imagePath, frame);
                    our_images.add(faceROI);
                }
                else if ((faces.empty())) {
                    System.out.println("--- No face detected ---");
                    break;
                }

                if (faceDetected) {
                    cap.release();
                    break;
                }
            } 

            long endTime = System.currentTimeMillis();
            long elapsedTime = endTime - startTime;
            System.out.println("Face detection time: " + elapsedTime + " ms");
            
            // Display the saved image
            HighGui.namedWindow("Detected face", HighGui.WINDOW_NORMAL);
            // Mat savedImage = Imgcodecs.imread(imagePath);
            //HighGui.imshow("Detected face", savedImage);
            
            // Release the camera and destroy the window
            cap.release();
            HighGui.destroyAllWindows();
            
            return faceDetected;
        } 
        catch (Exception e) {
            System.out.println("An error occurred: " + e.getMessage());
            return false;
        }
    }


    public static void binarization(Mat gray){
        // Perform simple thresholding
        int t = 120;                   // change threshold here
        int maxVal = 255;
        Mat binaryMask = new Mat();
        Imgproc.threshold(gray, binaryMask, t, maxVal, Imgproc.THRESH_BINARY);

        // Save the thresholded image
        Imgcodecs.imwrite("src/main/resources/DigitalAssistant/haarcascades/binaryMask.jpg", binaryMask);
    }


    public static void main(String[] args) {
        FaceDetection fd = new FaceDetection();
        System.out.println("starting face detection...");
        if (fd.faceDetection()) System.out.println("### Haar Cascades successfully applied ###");
        System.out.println("=== finished ===");
    }
}

File: src/main/python/FaceDetection.py
from py4j.clientserver import ClientServer, JavaParameters, PythonParameters

import numpy
import cv2


class FaceDetection(object):
    # This class handles object detection -> threshold based or Haar Cascade classifier (more sophisticated)

    def faceDetection(self):
        print("==== Hello there, the file is being executed. Hang on...")
        print("==== Starting face detection...")
        # Open the default camera (index 0)
        cap = cv2.VideoCapture(0)
        # Load the Haar cascade classifier for face detection
        face_cascade = cv2.CascadeClassifier(cv2.data.haarcascades + 'haarcascade_frontalface_default.xml')
        eye_cascade = cv2.CascadeClassifier(cv2.data.haarcascades + 'haarcascade_eye.xml')
        face_detected = False
        # Loop to continuously capture frames
        while True:
            # Read a new frame from the camera
            ret, frame = cap.read()

            # Convert the frame to grayscale for faster processing
            gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)

            # Detect faces in the frame using the Haar cascade classifier
            faces = face_cascade.detectMultiScale(gray, 1.3, 5)

            # Check if any faces were detected
            if len(faces) > 0:
                face_detected = True
                # Draw a rectangle around each face
                for (x, y, w, h) in faces:
                    cv2.rectangle(frame, (x, y), (x + w, y + h), (0, 255, 0), 5)
                    face_gray = gray[y:y + h, x:x + w]
                    face_color = frame[y:y + h, x:x + w]
                    # eye detection
                    eyes = eye_cascade.detectMultiScale(face_gray, 1.3, 5)
                    for (ex, ey, ew, eh) in eyes:
                        cv2.rectangle(face_color, (ex, ey), (ex + ew, ey + eh), (255, 0, 0), 5)

                # Display the frame with the detected faces in a window
                cv2.imshow('Object Detection', frame)

            # Check if the user pressed the 'q' key to quit
            if cv2.waitKey(1) == ord('q'):
                break

            # Check if the user pressed the 'esc' key to quit
            if cv2.waitKey(1) == 27:
                break
        print("Hello there...")
        # Release the camera and destroy the window
        cap.release()
        cv2.destroyAllWindows()

        print(face_detected)

        return True

    class Java:
        implements = ["DigitalAssistant.python.IFaceDetection"]


# Make sure that the python code is started first.
# Then execute: java -cp py4j.jar py4j.examples.SingleThreadClientApplication
facedetection = FaceDetection()
gateway = ClientServer(
    java_parameters=JavaParameters(),
    python_parameters=PythonParameters(),
    python_server_entry_point=facedetection)
